File: src/capability.ts
/**
 * Capability declaration parser for .mogdecl files.
 *
 * Parses declarations like:
 *   capability fs {
 *     fn read(path: string) -> string;
 *     fn write(path: string, content: string);
 *   }
 */

export interface CapabilityFuncDecl {
  name: string
  params: { name: string; type: string; optional: boolean }[]
  returnType: string | null // null = void
  isAsync: boolean
}

export interface CapabilityDecl {
  name: string
  functions: CapabilityFuncDecl[]
}

// Simple tokenizer for .mogdecl files
interface DeclToken {
  type: "keyword" | "ident" | "punct" | "eof"
  value: string
}

function tokenizeMogDecl(source: string): DeclToken[] {
  const tokens: DeclToken[] = []
  let i = 0

  while (i < source.length) {
    // Skip whitespace
    if (/\s/.test(source[i])) {
      i++
      continue
    }

    // Skip comments (// and #)
    if ((source[i] === "/" && source[i + 1] === "/") || source[i] === "#") {
      while (i < source.length && source[i] !== "\n") i++
      continue
    }

    // Punctuation
    if ("{};,()->?<>:".includes(source[i])) {
      // Handle -> as a single token
      if (source[i] === "-" && source[i + 1] === ">") {
        tokens.push({ type: "punct", value: "->" })
        i += 2
        continue
      }
      tokens.push({ type: "punct", value: source[i] })
      i++
      continue
    }

    // Identifiers / keywords
    if (/[a-zA-Z_]/.test(source[i])) {
      let start = i
      while (i < source.length && /[a-zA-Z0-9_]/.test(source[i])) i++
      const word = source.slice(start, i)
      const keywords = ["capability", "fn", "async", "Result"]
      tokens.push({
        type: keywords.includes(word) ? "keyword" : "ident",
        value: word,
      })
      continue
    }

    // Skip unknown characters
    i++
  }

  tokens.push({ type: "eof", value: "" })
  return tokens
}

class DeclParser {
  private tokens: DeclToken[]
  private pos = 0

  constructor(tokens: DeclToken[]) {
    this.tokens = tokens
  }

  private peek(): DeclToken {
    return this.tokens[this.pos] || { type: "eof", value: "" }
  }

  private advance(): DeclToken {
    const tok = this.tokens[this.pos]
    this.pos++
    return tok
  }

  private expect(type: string, value?: string): DeclToken {
    const tok = this.advance()
    if (tok.type !== type || (value !== undefined && tok.value !== value)) {
      throw new Error(
        `Expected ${type}${value ? ` '${value}'` : ""}, got ${tok.type} '${tok.value}'`,
      )
    }
    return tok
  }

  parse(): CapabilityDecl[] {
    const decls: CapabilityDecl[] = []
    while (this.peek().type !== "eof") {
      if (this.peek().value === "capability") {
        decls.push(this.parseCapability())
      } else {
        // Skip unknown tokens
        this.advance()
      }
    }
    return decls
  }

  private parseCapability(): CapabilityDecl {
    this.expect("keyword", "capability")
    const name = this.expect("ident").value
    this.expect("punct", "{")

    const functions: CapabilityFuncDecl[] = []
    while (this.peek().value !== "}") {
      if (this.peek().type === "eof") break
      functions.push(this.parseFunction())
    }

    this.expect("punct", "}")
    return { name, functions }
  }

  private parseFunction(): CapabilityFuncDecl {
    let isAsync = false
    if (this.peek().value === "async") {
      this.advance()
      isAsync = true
    }

    this.expect("keyword", "fn")
    const name = this.expect("ident").value
    this.expect("punct", "(")

    const params: { name: string; type: string; optional: boolean }[] = []
    while (this.peek().value !== ")") {
      if (params.length > 0) {
        this.expect("punct", ",")
      }
      const paramName = this.expect("ident").value
      this.expect("punct", ":")

      let optional = false
      let paramType = ""

      // Check for ?Type (optional)
      if (this.peek().value === "?") {
        this.advance()
        optional = true
      }

      paramType = this.parseTypeName()
      params.push({ name: paramName, type: paramType, optional })
    }

    this.expect("punct", ")")

    // Parse return type
    let returnType: string | null = null
    if (this.peek().value === "->") {
      this.advance()
      returnType = this.parseTypeName()
    }

    // Consume optional semicolon
    if (this.peek().value === ";") {
      this.advance()
    }

    return { name, params, returnType, isAsync }
  }

  private parseTypeName(): string {
    // Handle Result<T>
    if (this.peek().value === "Result") {
      this.advance()
      if (this.peek().value === "<") {
        this.advance()
        const inner = this.parseTypeName()
        this.expect("punct", ">")
        return `Result<${inner}>`
      }
      return "Result"
    }

    const name = this.expect("ident").value
    return name
  }
}

/**
 * Parse a .mogdecl file and return capability declarations.
 */
export function parseCapabilityDecl(source: string): CapabilityDecl[] {
  const tokens = tokenizeMogDecl(source)
  const parser = new DeclParser(tokens)
  return parser.parse()
}

/**
 * Parse a single capability from a .mogdecl file (convenience function).
 * Returns the first capability found, or null if none.
 */
export function parseCapability(source: string): CapabilityDecl | null {
  const decls = parseCapabilityDecl(source)
  return decls.length > 0 ? decls[0] : null
}

/**
 * Map a mogdecl type name to the Mog type system string.
 * Used during analysis to convert declaration types to the internal type representation.
 */
export function mogDeclTypeToMogType(declType: string): string {
  switch (declType) {
    case "int":
      return "i64"
    case "float":
      return "f64"
    case "bool":
      return "bool"
    case "string":
      return "string"
    case "none":
      return "void"
    default:
      // Handle Result<T> -> extract T
      if (declType.startsWith("Result<") && declType.endsWith(">")) {
        return mogDeclTypeToMogType(declType.slice(7, -1))
      }
      // Handle handle types - they become ptr
      return declType
  }
}

File: compiler/src/capability.rs
//! Capability declaration parser for `.mogdecl` files.
//!
//! Parses declarations like:
//! ```text
//! capability fs {
//!   fn read_file(path: string) -> string;
//!   async fn sleep(ms: int) -> int;
//! }
//! ```

use crate::types::Type;

// ── Public types ─────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct CapabilityDecl {
    pub name: String,
    pub functions: Vec<CapabilityFn>,
}

#[derive(Debug, Clone)]
pub struct CapabilityFn {
    pub name: String,
    pub params: Vec<(String, Type)>,
    pub return_type: Type,
    pub is_async: bool,
}

// ── Tokenizer ────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
enum TokKind {
    Keyword,
    Ident,
    Punct,
    Eof,
}

#[derive(Debug, Clone)]
struct Tok {
    kind: TokKind,
    value: String,
}

fn tokenize_mogdecl(source: &str) -> Vec<Tok> {
    let src: Vec<char> = source.chars().collect();
    let mut tokens = Vec::new();
    let mut i = 0;

    while i < src.len() {
        // Skip whitespace
        if src[i].is_ascii_whitespace() {
            i += 1;
            continue;
        }

        // Skip comments (// and #)
        if (src[i] == '/' && i + 1 < src.len() && src[i + 1] == '/') || src[i] == '#' {
            while i < src.len() && src[i] != '\n' {
                i += 1;
            }
            continue;
        }

        // Punctuation
        if "{};,()->?<>:".contains(src[i]) {
            // Handle -> as a single token
            if src[i] == '-' && i + 1 < src.len() && src[i + 1] == '>' {
                tokens.push(Tok {
                    kind: TokKind::Punct,
                    value: "->".into(),
                });
                i += 2;
                continue;
            }
            tokens.push(Tok {
                kind: TokKind::Punct,
                value: src[i].to_string(),
            });
            i += 1;
            continue;
        }

        // Identifiers / keywords
        if src[i].is_ascii_alphabetic() || src[i] == '_' {
            let start = i;
            while i < src.len() && (src[i].is_ascii_alphanumeric() || src[i] == '_') {
                i += 1;
            }
            let word: String = src[start..i].iter().collect();
            let kind = match word.as_str() {
                "capability" | "fn" | "async" | "Result" => TokKind::Keyword,
                _ => TokKind::Ident,
            };
            tokens.push(Tok { kind, value: word });
            continue;
        }

        // Skip unknown characters
        i += 1;
    }

    tokens.push(Tok {
        kind: TokKind::Eof,
        value: String::new(),
    });
    tokens
}

// ── Parser ───────────────────────────────────────────────────────────

struct DeclParser {
    tokens: Vec<Tok>,
    pos: usize,
    eof: Tok,
}

impl DeclParser {
    fn new(tokens: Vec<Tok>) -> Self {
        Self {
            tokens,
            pos: 0,
            eof: Tok {
                kind: TokKind::Eof,
                value: String::new(),
            },
        }
    }

    fn peek(&self) -> &Tok {
        self.tokens.get(self.pos).unwrap_or(&self.eof)
    }

    fn advance(&mut self) -> Tok {
        let tok = self.tokens.get(self.pos).cloned().unwrap_or(Tok {
            kind: TokKind::Eof,
            value: String::new(),
        });
        self.pos += 1;
        tok
    }

    fn expect(&mut self, kind: TokKind, value: Option<&str>) -> Result<Tok, String> {
        let tok = self.advance();
        if tok.kind != kind {
            return Err(format!(
                "Expected {:?}{}, got {:?} '{}'",
                kind,
                value.map_or(String::new(), |v| format!(" '{}'", v)),
                tok.kind,
                tok.value,
            ));
        }
        if let Some(v) = value {
            if tok.value != v {
                return Err(format!("Expected '{}', got '{}'", v, tok.value));
            }
        }
        Ok(tok)
    }

    fn parse(&mut self) -> Result<Vec<CapabilityDecl>, String> {
        let mut decls = Vec::new();
        while self.peek().kind != TokKind::Eof {
            if self.peek().value == "capability" {
                decls.push(self.parse_capability()?);
            } else {
                self.advance();
            }
        }
        Ok(decls)
    }

    fn parse_capability(&mut self) -> Result<CapabilityDecl, String> {
        self.expect(TokKind::Keyword, Some("capability"))?;
        let name = self.expect(TokKind::Ident, None)?.value;
        self.expect(TokKind::Punct, Some("{"))?;

        let mut functions = Vec::new();
        while self.peek().value != "}" {
            if self.peek().kind == TokKind::Eof {
                break;
            }
            functions.push(self.parse_function()?);
        }

        self.expect(TokKind::Punct, Some("}"))?;
        Ok(CapabilityDecl { name, functions })
    }

    fn parse_function(&mut self) -> Result<CapabilityFn, String> {
        let mut is_async = false;
        if self.peek().value == "async" {
            self.advance();
            is_async = true;
        }

        self.expect(TokKind::Keyword, Some("fn"))?;
        let name = self.expect(TokKind::Ident, None)?.value;
        self.expect(TokKind::Punct, Some("("))?;

        let mut params = Vec::new();
        while self.peek().value != ")" {
            if !params.is_empty() {
                self.expect(TokKind::Punct, Some(","))?;
            }
            let param_name = self.expect(TokKind::Ident, None)?.value;
            self.expect(TokKind::Punct, Some(":"))?;

            // Check for ?Type (optional) — consume the ? but note it
            if self.peek().value == "?" {
                self.advance();
            }

            let param_type = self.parse_type_name()?;
            params.push((param_name, param_type));
        }

        self.expect(TokKind::Punct, Some(")"))?;

        // Return type
        let return_type = if self.peek().value == "->" {
            self.advance();
            self.parse_type_name()?
        } else {
            Type::Void
        };

        // Consume optional semicolon
        if self.peek().value == ";" {
            self.advance();
        }

        Ok(CapabilityFn {
            name,
            params,
            return_type,
            is_async,
        })
    }

    fn parse_type_name(&mut self) -> Result<Type, String> {
        // Handle Result<T>
        if self.peek().value == "Result" {
            self.advance();
            if self.peek().value == "<" {
                self.advance();
                let inner = self.parse_type_name()?;
                self.expect(TokKind::Punct, Some(">"))?;
                return Ok(Type::Result(Box::new(inner)));
            }
            return Ok(Type::Result(Box::new(Type::Void)));
        }

        let tok = self.advance();
        Ok(mogdecl_type_to_type(&tok.value))
    }
}

// ── Type mapping ─────────────────────────────────────────────────────

/// Convert a mogdecl type name to the Mog `Type`.
pub fn mogdecl_type_to_type(name: &str) -> Type {
    match name {
        "int" => Type::int(),
        "float" => Type::float(),
        "bool" => Type::Bool,
        "string" => Type::String,
        "none" => Type::Void,
        other => {
            // Handle Result<T> string form
            if let Some(inner) = other
                .strip_prefix("Result<")
                .and_then(|s| s.strip_suffix('>'))
            {
                return Type::Result(Box::new(mogdecl_type_to_type(inner)));
            }
            Type::Custom(other.to_string())
        }
    }
}

// ── Public API ───────────────────────────────────────────────────────

/// Parse a `.mogdecl` file and return all capability declarations found.
pub fn parse_capability_decls(content: &str) -> Result<Vec<CapabilityDecl>, String> {
    let tokens = tokenize_mogdecl(content);
    let mut parser = DeclParser::new(tokens);
    parser.parse()
}

/// Parse a `.mogdecl` file and return the first capability, or `None`.
pub fn parse_capability_decl(content: &str) -> Option<CapabilityDecl> {
    let decls = parse_capability_decls(content).ok()?;
    decls.into_iter().next()
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_process_capability() {
        let src = r#"
            capability process {
                async fn sleep(ms: int) -> int;
                fn getenv(name: string) -> string;
                fn exit(code: int) -> int;
            }
        "#;
        let decl = parse_capability_decl(src).expect("should parse");
        assert_eq!(decl.name, "process");
        assert_eq!(decl.functions.len(), 3);

        assert_eq!(decl.functions[0].name, "sleep");
        assert!(decl.functions[0].is_async);
        assert_eq!(decl.functions[0].params.len(), 1);
        assert_eq!(decl.functions[0].params[0].0, "ms");
        assert_eq!(decl.functions[0].return_type, Type::int());

        assert_eq!(decl.functions[1].name, "getenv");
        assert!(!decl.functions[1].is_async);
        assert_eq!(decl.functions[1].return_type, Type::String);
    }

    #[test]
    fn parse_log_capability_no_return_types() {
        let src = r#"
            capability log {
                fn info(message: string);
                fn warn(message: string);
            }
        "#;
        let decl = parse_capability_decl(src).expect("should parse");
        assert_eq!(decl.name, "log");
        assert_eq!(decl.functions.len(), 2);
        assert_eq!(decl.functions[0].return_type, Type::Void);
    }

    #[test]
    fn parse_with_comments() {
        let src = r#"
            // This is a comment
            # Another comment style
            capability http {
                async fn get(url: string) -> string;
            }
        "#;
        let decl = parse_capability_decl(src).expect("should parse");
        assert_eq!(decl.name, "http");
        assert_eq!(decl.functions.len(), 1);
    }

    #[test]
    fn empty_input_returns_none() {
        assert!(parse_capability_decl("").is_none());
        assert!(parse_capability_decl("// just a comment").is_none());
    }

    #[test]
    fn mogdecl_type_mapping() {
        assert_eq!(mogdecl_type_to_type("int"), Type::int());
        assert_eq!(mogdecl_type_to_type("float"), Type::float());
        assert_eq!(mogdecl_type_to_type("bool"), Type::Bool);
        assert_eq!(mogdecl_type_to_type("string"), Type::String);
        assert_eq!(mogdecl_type_to_type("none"), Type::Void);
        assert_eq!(
            mogdecl_type_to_type("Result<int>"),
            Type::Result(Box::new(Type::int()))
        );
    }
}
